### PR TITLE
Added first version dotnet core 3.0 template

### LINKED ISSUE #176: Dotnet template naming and dotnet core 2.2 and 3.0 templates
We need the option to select 2.2 and 3.0 dotnet core template. We also need to address the naming for these templates. I think using "csharp" as template name is misleading.

### PR DESCRIPTION
**Pull request by csakshaug**
dotnet core 3.0 template

## Description
This a is dotnet core 3.0 template based on the csharp template, but it also updated the base docker image location (using new Microsoft registry) and also fixed the that the handler uses dotnet core and not standard.

_The csharp template has mixed with net core and standard 2.1. It also has some flaws from a "dotnet developing perspective", I have not addressed this in this pull request._

## Motivation and Context
We should have updated templates both for 2.2 and 3.0. The name csharp could be misleading, my suggestion is to start using "netcore" or "netstandard" as basename instead of "csharp".

The reason why I have selected netcore is same as Microsoft using as a prefix for the framework in the csproj file.

- [x] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))  

## Which issue(s) this PR fixes 
Fixes #176

## How Has This Been Tested?
Only testet docker build

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Version change (see: Impact to existing users)

## Impact to existing users
None, new naming for dotnet core templates

## Checklist:
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [ ] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by alexellis** [comment]
I suspect that .NET 3.0 is a good time for us to move to a HTTP server and of-watchdog

How about using Kestrel?

Template name `netcore3-http`?

**Comment by csakshaug** [comment]
> I suspect that .NET 3.0 is a good time for us to move to a HTTP server and of-watchdog

I agree

> How about using Kestrel?

I do that in other projects

> Template name `dotnetcore3-http`?

dotnetcore3-http for this one? another with kestrel etc?

**Comment by alexellis** [comment]
Instead of adding a STDIO template for .NET Core 3.0, let's only add a http-based template using Kestrel.

File: template/netcore30/Program.cs [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+using System.Text;
+using Function;
+
+namespace root
+{
+    class Program
+    {
+        static void Main(string[] args)
+        {
+            string buffer = Console.In.ReadToEnd();
```
Comment (michaelkacher) [inline template/netcore30/Program.cs]: Have you considered changing to to ```static async Task Main(string[] args)``` and changing to ```await Console.In.ReadToEndAsync();```?  If you do this you could also change the FunctionHandler to be async.

Comment (csakshaug) [reply template/netcore30/Program.cs]: I copy the code, but I think it some potentials to optimize the implementations before the migrations. Should we consider to append different templates like we talking on slack? -http, -kestrel, -async?

**Comment by csakshaug** [comment]
What are we thinking about the name?

1. netcore vs dotnetcore vs aspnetdotnetcore
2. -http, -kestrel etc
3. I see that in other templates we operate with major version number, how can we have support for both 2.0,2.1,2.2,3.0, 3.1 with only numbers?

**Comment by burtonr** [comment]
The overall proposal conversation should occur on the issue as mentioned in the [contribution guide](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md#i-have-a-great-idea)

The conversations in the PR should be focused around the implementation details.

**Comment by alexellis** [comment]
Various options:

* `csharp-`
* `dncore-`
* `dotnetcore-`
* `netcore-`

**Comment by alexellis** [comment]
`csharp-dotnetcore3.0`
`csharp-netcore3.0`

`csharp-dotnetcore3.0-kestrel`
`csharp-netcore3.0-kestrel`

`csharp-3.0-kestrel`

**Comment by alexellis** [comment]
For anyone landing here, please see: https://www.openfaas.com/blog/asp-net-core/

**Comment by alexellis** [comment]
I am closing out these tickets, but want the work to continue, as an of-watchdog template

**Pull request closed without merging**